### PR TITLE
SBRA-571 : document the good practice for new code

### PR DESCRIPTION
**Pull request by A669015**
This PR re-write the main README file to give the guidelines to experiment the existing model projects and how to contribute to existing one or create a new one.

The model project specific README have been initiated, but will be filled in a second step by the specialists of the different use-cases. Additional JIRA tasks have been created for that.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+* ``data.py`` contains the dataset and datamodule.
+* ``models.py`` contains the model ("module"), including training logic. The architecture can be imported from specific librairies or a local module (e.g. unet.py).
+* ``trainer.py`` is the main entrypoint responsible for creating a Trainer object, a CLI, and saving artifacts in the experiment directory.
+* ``noxfile.py`` is the Nox build tool configuration file that defines all targets available for the model project.
+* ``env.yaml`` lists the environment requirements, like the Ubuntu docker base image or the Python version.
+* ``requirements.txt`` contains all the python dependencies of the project, generated using the ``pip freeze`` command.
+* ``ci/`` contains all the additional files required to run the functional tests. 
+
+Optionally it can also include the following folders and files :
+* ``data/`` contains the dataset configuartaion files and will contain the `raw` and `processed` data directories, and normalization factors (and optionally explicit train/val/test split sets), once the dataset as been downloaded and processed.
```
Comment (mikael10j) [inline README.md:74]: configuration

File: README.md [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+nox==2024.4.15
+torch==1.13.1+cu117
+pytorch-lightning==1.5.7
+```
+
+Note this is not the final version of this file, because you will update it at the end of the process (see [Finalize the requirements.txt](#finalize-the-requirementstxt)), just before pushing your development branch to create a pull request. As a concequence, you can add other requirements right now, but any python library you will install, using ``pip install``, during your development process, will be included in the final version of the file.
+
+##### Build the development environment
+
+You are now ready to setup your development environment, either a *Docker container* or a Python *virtual environment*.
```
Comment (mikael10j) [inline README.md:329]: either as a...

File: README.md [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+To build your development docker image, use the generic ``Dockerfile`` in the ``docker`` folder.
+
+The 4 following parameters must be passed to the Dockerfile, the values being picked, for the 3 last ones, from the env.yaml file previously created :
+- MODEL_PROJECT_PATH : the path ``<domain>/<use-case>/<NN architecture>`` you are working on.
+- UBUNTU_IMAGE_NAME : the base Ubuntu docker image name indicated in the ``env.yaml`` file.
+- UBUNTU_IMAGE_TAG : the base Ubuntu docker image tag indicated in the ``env.yaml`` file.
+- PYTHON_VERS : the Python version indicated in the ``env.yaml`` file.
+
+To build the Docker image run from the repository root directory :
```
Comment (mikael10j) [inline README.md:341]: I think you can remove "run"

File: README.md [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+You are now ready to setup your development environment, either a *Docker container* or a Python *virtual environment*.
+
+###### Docker container
+
+To build your development docker image, use the generic ``Dockerfile`` in the ``docker`` folder.
+
+The 4 following parameters must be passed to the Dockerfile, the values being picked, for the 3 last ones, from the env.yaml file previously created :
+- MODEL_PROJECT_PATH : the path ``<domain>/<use-case>/<NN architecture>`` you are working on.
+- UBUNTU_IMAGE_NAME : the base Ubuntu docker image name indicated in the ``env.yaml`` file.
```
Comment (mikael10j) [inline README.md:337]: I'm not sure to understand. Do we need to write it explicitely or will it be automatically retrieved from the env.yaml?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    -v <domain>/<use-case>/<NN architecture>:/home/ai4sim/<domain>/<use-case>/<NN architecture> \
+    model-collection:<domain>-<use-case>-<NN architecture> \
+    bash
+```
+
+In the container, you can directly use the ``pip`` and ``python`` command of the system to respectively install the dependencies and execute your code.
+
+###### Virtual environment
+
+To build your virtual environment, please be sure to have the Python version, indicated in our env.yaml file, installed on your system, and the *Nox* tool available has described [previously](#installation).
```
Comment (mikael10j) [inline README.md:366]: Python version indicated...

Comment (mikael10j) [reply README.md:366]: in your env.yaml file installed...

File: README.md [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+##### Linting
+
+The linting is a static code analysis in charge of checking the code formating to avoid some basic bugs and ensure an homogeneity of the code base with respect of the main Python checkstyle rules. We use [flake8](https://flake8.pycqa.org/en/latest/) and some plugins to rate the code.
+
+To evalute your checkstyle code quality run :
+```
+nox -s lint -v
+```
+
+The rating score must be 10/10, so all checkstyle violationreturned by the above command should be fixed.
```
Comment (mikael10j) [inline README.md:392]: violation returned...

File: README.md [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+
+Nevertheless, if after all your best effort, you think a checkstyle violation cannot/should not be fixed, you can deactivate the lint checker on a given line of code with the [noqa](https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html#in-line-ignoring-errors) directive:
+```
+example = lambda: 'example'  # noqa: E731
+```
+
+##### Unit tests
+
+The new code should be provided with as much as possible unit tests. Because the main part of the code is based on *lightning* and *torch*, it is not interresting to test all classes and method inherited from these libraries, but there is still some part of code interresting to be tested.
```
Comment (mikael10j) [inline README.md:401]: as many as possible(?)

Comment (mikael10j) [reply README.md:401]: interesting

Comment (mikael10j) [reply README.md:401]: and methods

File: README.md [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+
+```
+nos -s tests -v
+```
+
+##### Functional tests
+
+Each model project should embed functional tests in the ``ci`` folder, that ensure the training code is purely functional, i.e. without any precision or performance consideration.
+
+The principle of this functional test is to launch a training on a dummy dataset. To be sure to detect possible regression due to changes in the code, and not on the data source side, the dummy dataset must be generated synthetically.
```
Comment (mikael10j) [inline README.md:415]: regressions or a possible regression

File: README.md [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+nos -s tests -v
+```
+
+##### Functional tests
+
+Each model project should embed functional tests in the ``ci`` folder, that ensure the training code is purely functional, i.e. without any precision or performance consideration.
+
+The principle of this functional test is to launch a training on a dummy dataset. To be sure to detect possible regression due to changes in the code, and not on the data source side, the dummy dataset must be generated synthetically. 
+
+The ``ci`` folder must contains
```
Comment (mikael10j) [inline README.md:417]: contain

File: README.md [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+```
+
+##### Functional tests
+
+Each model project should embed functional tests in the ``ci`` folder, that ensure the training code is purely functional, i.e. without any precision or performance consideration.
+
+The principle of this functional test is to launch a training on a dummy dataset. To be sure to detect possible regression due to changes in the code, and not on the data source side, the dummy dataset must be generated synthetically. 
+
+The ``ci`` folder must contains 
+- a ``generate_synthetic_data.py`` file in charge of the synthetic dataset generation based on the real data format. This script must taking care of not erasing real data already present in the ``data`` folder. We sugges to raise an exception with an explicit message in case of presence of existing data in the targetted path. Future work, should address this issue making the data path configurable.
```
Comment (mikael10j) [inline README.md:418]: must take care
We suggest
targeted
Future work should...

File: README.md [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+- a ``configs`` folder with the ligthning formated configuration file of the test training,
+- a ``run.sh`` script file in charge of running the different training tests.
+
+Optionally it can alos includes :
```
Comment (mikael10j) [inline README.md:422]: also include

File: README.md [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+- a ``requirements_data.txt`` listing the dependencies related to the synthetic data generation.
+
+The *Nox* session ``train_test`` allows to run the functional tests :
+```
+nox -s train_test -v
+```
+
+By default, the synthetic data are kept in the ``data`` folder, after the tests had run. To clean the ``data`` folder at the end of the tests, use the ``clean_data`` option :
```
Comment (mikael10j) [inline README.md:430]: the tests have run.

File: README.md [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+```
+nox -s train_test -v -- clean_data
+```
+
+### Pull request
+
+Now, your development are done, the last step is to open a *Pull Request* to propose your contribution.
```
Comment (mikael10j) [inline README.md:437]: your developments are

File: README.md [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+
+#### Document the model project
+
+In each model project a ``README.md`` file should :
+
+- introduce the scientific challenge adressed,
+- how to get the dataset,
```
Comment (mikael10j) [inline README.md:444]: explain how to...

File: README.md [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+```
+pip freeze > requirements.txt
+```
+
+Then commit this last change.
+
+#### Add the model project to the CI/CD workflows
+
+The CI/CD workflows are defined by the Github Actions Yaml files in the ``workflows`` folder. There are 2 different workflow :
+- the *code quality* CI/CD workflow : this workflow (see ``workflows/ai4sim-ci-cd.yaml``) is in charge of running tests and lint for all model project, to detect possible regressions. It is triggered on any changed done on a any branch and on any pull requests.
```
Comment (mikael10j) [inline README.md:461]: for all model projects ... it is triggered by any change done on any branch and on any pull request.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+- the *docker image builder* CI/CD workflow : this workflow (see ``workflows/images-ci-cd.yaml``) is in charge of building and publishing the docker images for all model projects. It is triggered on any changed done on the *main* branch and on any pull requests targetting the *main* branch.
```
Comment (mikael10j) [inline README.md:462]: It is triggered by any change done on the... any pulll request targeting...

File: README.md [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+
+The CI/CD workflows run the same actions for all model projects listed in the *jobs/strategy/matrix/model-project* section of the Yaml files.
+
+To finalize your contribution you than have to add the path of your model project in all the CI/CI workflow files :
```
Comment (mikael10j) [inline README.md:466]: you then have...

File: README.md [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+          - weather-forecast/ecrad-3d-correction/unets
+          - <domain>/<use-case>/<NN architecture>       <-- Your new model project path
+...
+```
+
+Then, commit these last changes and push your branch to github. It should trigger the CI/CD workflow related to the *code quality*. If all the CI/CD runs pass, you are now able to open a *Pull Request*.
+
+#### Open the *Pull Request*
+
+ Open a *Pull Request* to merge your development branch in the *main* branch, with a description of your contribution as much as possible complete.
```
Comment (mikael10j) [inline README.md:486]: branch into the *main*

File: README.md [hunk 19]
```diff
@@ -0,0 +1,492 @@
...
+
+Using the docker image described at [Setting-up the environment](#setting-up-the-environment), you can launch the training of the model with :
+
+```
+cd <domain>/<use-case>/<NN architecture>
+podman run \
+    -v ./data:/home/ai4sim/<domain>/<use-case>/<NN architecture>/data \
+    -w /home/ai4sim/<domain>/<use-case>/<NN architecture> \
+    ghcr.io/ai4sim/model-collection:<domain>-<use-case>-<NN architecture> \
+    python3 trainer.py --config configs/<training-config>.yaml
```
[the remark this answers is not in the document]
Comment (A669015) [reply README.md:163]: Yes we can change the default behavior, now, but I have to validate it before updating the documentiion.

File: README.md [hunk 20]
```diff
@@ -0,0 +1,492 @@
...
+This file will actually be used later by the github actions workflow in charge of building the docker image of the model project, but we recommend to create and fill it now to fix :
+- the **Ubuntu image** you will work on,
+- the **Python version** you want to use.
+
+The format of this file is the following :
+```
+# This file defines the global environment dependencies for this model project.
+# It is read by the CI/CD workflows in charge of the tests (ai4sim-ci-cd.yaml)
+# and the build of the Docker image (images-ci-cd.yaml) to set the environment.
+python_version: 3.8
```
[the remark this answers is not in the document]
Comment (A669015) [reply README.md:262]: I don't think it will cause conflict, the python version described in the env.yaml file will be installed and used as new default pyhton3. But the "old" one, and its dependencies will no more be used.

In this case (of a pre-configured image), we could document that the python version described in the env.yaml file **must be** the one already available in the docker image to take part of the pre-installed python libs. Is it satisfying enough ?

We could also imagine to make the python version field optional in the env.yaml file, but it will require to manage the case of images without pre-installed python. Maybe it could be adressed later to improve the UX, if needed ?

File: README.md (hunk 4 above)
Comment (A669015) [reply README.md:337]: It should be written explicitely using the  --build-arg docker option, as descibed in the example below. I will modify to insist on that.

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (mikael10j) [reply README.md:163]: Ok, let it as it is for the moment, and we'll change in a future update

File: README.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (mikael10j) [reply README.md:262]: For the moment, additional info in case of images including python is ok for me. Let's discuss later potential improvements.

**Review by mikael10j — APPROVED** [review]

**Review by elasto — APPROVED** [review]

File: README.md [hunk 21]
```diff
@@ -1,61 +1,73 @@
...
+``<domain>``/``<use-case>``/``<NN architecture>``
+
+For example, the code in the path ``cfd/combustion/gnns``, implements some **Graph Neural Network (GNN)** achitectures developed for the **Computational Fluid Dynamics (CFD)**, and applied to a **Combustion** use-case.
 
-* *data/raw* stores the raw data files;
-* *data/processed* stores the data after preprocessing, used for several experiments.
+### Model project files
 
-## Collections
+To take care of the boiler-plate, early stopping, tensorboard logging, training parallelization etc. we integrate directly with PyTorch Lightning. For each new  model, we use the same code architecture, described below.
```
Comment (elasto) [inline README.md]: we integrate them directly with PyTorch Lightning

File: README.md [hunk 22]
```diff
@@ -1,61 +1,73 @@
...
-* Weather Forecast
+### Projects architecture
 
-    - Gravity Wave Drag
-    - 3D Bias Correction (with Unets and Attention CNNs)
+Each model project is made of two pipelines:
+
+* Data pipeline: a **DataModule** wraps a Dataset to provide data (with dataloaders, preprocessing...);
+* Model pipeline: a **Module** compiles a neural network architecture with its optimizer.
+The Trainer plugs both pipelines when called to run an experiment. It can be called by hand (CLI, with config files) or by another abstraction layer (e.g. hyperparameters optimization, meta learning)0
```
Comment (elasto) [inline README.md]: (e.g. hyperparameters optimization, meta learning)

File: README.md [hunk 23]
```diff
@@ -65,27 +87,92 @@ The Dataset can be implemented with various librairies, following the implicit c
...
 
-A docker file is provided to get started. To build the image, run the following command:
-::
-    docker build -t ai4sim .
+#### Docker container
 
-To install the corresponding python requirements for a use-case an entrypoint is implemented. It can be selected by one of the folling [combustion_gnns, combustion_unets, wf_gwd]:
-::
-    docker run -it --rm ai4sim ./script.sh combustion_gnns
+The AI4SIM github CI/CD publishes in the [github registry](https://github.com/AI4SIM/model-collection/pkgs/container/model-collection) a docker image dedicated to each model project proposed in the **Model collection** repository. Each image is built on an public **Ubuntu** base image (e.g. *nvidia/cuda:11.7.1-cudnn8-runtime-ubuntu20.04*). All the model project's requirements have been installed and the model project code has been added in ``/home/ai4sim/<domain>/<use-case>/<NN architecture>``.
```
Comment (elasto) [inline README.md]: The model project code has been added in ``/home/ai4sim/<domain>/<use-case>/<NN architecture>`` and its requirements have been installed in the container

File: README.md [hunk 24]
```diff
@@ -168,40 +171,80 @@ python3 trainer.py --config configs/<training-config>.yaml
...
+
+#### Create the proper python development environment
 
 The nox target are also very useful to launch generic command during development phase.
 
-#### Run unit tests
+##### Run unit tests
 
 You can run the whole unit test suite of a use case, using ``pytest``, with ``nox -s tests``.
 This target also prints out the coverage report and save a xml version in ``.ci-reports/``.
```
Comment (elasto) [inline README.md]: This target also prints out the coverage report and save it in a xml version in ``.ci-reports/``.

File: README.md [hunk 25]
```diff
@@ -168,40 +171,80 @@ python3 trainer.py --config configs/<training-config>.yaml
...
 
-#### Run linting
+##### Run linting
+
+You can run the python linting of the code use case, using ``flake8``, with ``nox -s lint``.
+
+
+### Existing model project
+
+Please note, the maintenance policy of the repository is to consider a model project freezed from the moment when it is intergrated. It means you can open an issue, but the only major ones (bugs breaking the functionality or leading the model to be scientifically unrelevant) will be treated, but no upgrade nor refactoring of the existing model code will be adressed.
```
Comment (elasto) [inline README.md]: but neither the upgrade nor the refactoring of the existing model code will be adressed.

**Review by elasto — CHANGES_REQUESTED** [review]
I just reformulate some phrase or fix some typos

File: README.md [hunk 26]
```diff
@@ -0,0 +1,498 @@
...
+The collection of models is developed through partnerships bringing their use-cases.
+
+Currently, the models that have been developed are based on the following use-cases:
+
+- Computational Fluid Dynamics
+    - Combustion (with UNets and GNNs), from CERFACS
+
+- Weather Forecast
+    - Gravity Wave Drag (with CNNs), from ECMWF
+    - 3D Bias Correction (with UNets), from ECMWF
```
Comment (cbovalo) [inline README.md:40]: 3D cloud radiative effect correction

File: README.md [hunk 27]
```diff
@@ -0,0 +1,498 @@
...
+* ``ci/`` contains all the additional files required to run the functional tests.
+
+Optionally, it can also include the following folders and files:
+* ``data/`` contains the dataset configuration files and will contain the `raw` and `processed` data directories, and normalization factors (and optionally explicit train/val/test split sets), once the dataset has been downloaded and processed.
+* ``notebooks/`` contains example Jupyter notebooks, for pedagogical purposes.
+* ``plotters.py`` takes care of plot generation for the test set.
+
+### Project Architecture
+
+Each model project is made of two pipelines:
```
Comment (cbovalo) [inline README.md:78]: Maybe specify that the two pipelines come from Lightning

**Review by elasto — APPROVED** [review]
